_posts/2020-07-29-springboot-logging.md

```diff
@@ -0,0 +1,336 @@
+Logs become even more useful if we add some contextual information to every message we log, for example, the identifier of the user or the name of the business function. Instead of manually appending the contextual information to each log message, we can use Logback’s Mapped Diagnostic Context (MDC). Whatever we put in the MDC can be used in the log pattern. This behavior is comparable to a ThreadLocal variable.
+For example, we put the user idenntifier and function name in the MDC for adding to the log message:
+
+```java
+  @GetMapping("/users/{userID}")
+  public String getUser(@PathVariable("userID") final String userID) {
+    
+    MDC.put("user", userID);
+    MDC.put("function", "userInquiry");
+    
+    logger.info("Controller: Fetching user with id {}", userID);
+    
+    MDC.remove("user");
+    MDC.remove("function");
+    
+    return userService.getUser(userID);
+                                                 
+```
+
+The user identifier and function name in the MDC are then added to the log message in the pattern attribute :
+```
+%d{HH:mm:ss.SSS} [%X{user}] [%X{function}] [%thread] %-5level %logger{36} - %msg%n
+```
+A sample of the log output with the MDC attributes user and function name :
+```shell
+...            : Controller: Fetching user with id john123
+... [john123] [userInquiry] ... - Controller: Fetching user with id john123
+...       : Service: Fetching user with id john123
+... [john123] [userInquiry] ... - Service: Fetching user with id john123
+```
+
+### Different Logging For Each Environment
+
+We often have different logging formats for local and production runtime environments. Spring profiles are an elegant way to implement different logging for each environment. You can refer to a very good use case in this article for (environment-specific logging)[https://reflectoring.io/profile-specific-logging-spring-boot/]. 
+
+### Tracing Requests Across Microservices
+Debugging and tracing in microservice applications is challenging since
+the microservices are deployed and run independently resulting in their logs being distributed across many individual components. We apply two techniques to help simplify the log analysis process of multiple microservices : 
+
+#### Log Aggregation 
+Logs from different microservices are aggregated to a central location. For spring boot, we need to output logs in a format compatible with the log aggregation software. Let us look at an appender configured for logstash :
+
+```xml
+  <appender name="LOGSTASH"
+    class="net.logstash.logback.appender.LogstashTcpSocketAppender">
+    <destination>localhost:4560</destination>
+    <encoder charset="UTF-8"
+      class="net.logstash.logback.encoder.LogstashEncoder" />
+  </appender>
+``` 
+Here the LogstashEncoder encodes logs in JSON format and sent to the elastic search database. We can then apply various visualization tools to query logs.
+
+
+#### Correlate Logs To Trace Requests Across Microservices
```

Comment:
Remove this section and just declare that it's possible and link to the other article in the previous section.